x/btccheckpoint/keeper/msg_server.go

```diff
@@ -47,25 +49,40 @@ func (ms msgServer) InsertBTCSpvProof(ctx context.Context, req *types.MsgInsertB
 		return nil, types.ErrInvalidHeader.Wrap(err.Error())
 	}
 
+	epochNum := rawSubmission.CheckpointData.Epoch
+
+	ed := ms.k.GetEpochData(sdkCtx, epochNum)
+
+	if ed == nil {
+		// we do not have any data saved yet
+		newEd := types.NewEmptyEpochData()
+		ed = &newEd
+	}
+
+	if ed.Status == types.Finalized {
+		// we have already finalized given epoch so we do not need any more submissions
+		return nil, types.ErrEpochAlreadyFinalized
+	}
+
 	// At this point:
 	// - every proof of inclusion is valid i.e every transaction is proved to be
 	// part of provided block and contains some OP_RETURN data
 	// - header is proved to be part of the chain we know about through BTCLightClient
+	// - epoch is not yet finalized
 	// - this is new checkpoint submission
 	// Verify if this is expected checkpoint
-	err = ms.k.checkpointingKeeper.VerifyCheckpoint(sdkCtx, rawSubmission.CheckpointData)
+	if err := ms.k.checkpointingKeeper.VerifyCheckpoint(sdkCtx, rawSubmission.CheckpointData); err != nil {
+		if errors.Is(err, ckpttypes.ErrConflictingCheckpoint) {
+			// We end such transaction with success to preserve setting of the conflict
+			// flag in the state. This flag will trigger halt of the chain in the
+			// EndBlocker call
```

Comment:
Is this part of the pr?

Reply:
this is already part of checkpointing module. It was done in this pr https://github.com/babylonlabs-io/babylon/pull/530